src/ui/MissingTerm/MissingTerm.ts

```diff
@@ -0,0 +1,107 @@
+import 'styling/_MissingTerm';
+import { exportGlobally } from '../../GlobalExports';
+import { Component } from '../Base/Component';
+import { IComponentBindings } from '../Base/ComponentBindings';
+import { ComponentOptions } from '../Base/ComponentOptions';
+import { reject } from 'underscore';
+import { $$, Initialization } from '../../Core';
+import { Dom } from '../../utils/Dom';
+import { analyticsActionCauseList, IAnalyticsIncludeMissingTerm } from '../Analytics/AnalyticsActionListMeta';
+
+export interface IMissingTermOptions {
+  caption?: string;
+  clickable?: boolean;
+}
+
+export class MissingTerm extends Component {
+  static ID = 'MissingTerm';
+  static options: IMissingTermOptions = {
+    /**
+     * Enable the user to click on a missing term to force it in the query
+     */
+    clickable: ComponentOptions.buildBooleanOption({ defaultValue: true }),
+    /**
+     * Change the text that appears before the missing keywords
+     */
+    caption: ComponentOptions.buildLocalizedStringOption({ defaultValue: 'Missing' })
+  };
+
+  static doExport = () => {
+    exportGlobally({
+      MissingTerm: MissingTerm
+    });
+  };
+  /**
+   * The `MissingTerm` component is responsible for displaying the keyword that was not matched for a result
+   */
+  constructor(public element: HTMLElement, public options?: IMissingTermOptions, bindings?: IComponentBindings, public result?) {
+    super(element, MissingTerm.ID, bindings);
+
+    this.options = ComponentOptions.initComponentOptions(element, MissingTerm, options);
+    this.addMissingTerm();
+  }
+
+  private addMissingTerm() {
+    if (this.getMissingTerms().length > 0) {
+      const container = this.buildContainer();
+      $$(this.element).append(container.el);
+      const cell = $$(this.element).find('.coveo-result-cell');
+      const caption = this.buildCaption();
+      cell.appendChild(caption.el);
+      this.buildMissingTerms().forEach(term => {
+        cell.appendChild(term.el);
+      });
+    }
+  }
+
+  private buildContainer(): Dom {
+    return $$('div', { className: 'coveo-result-row' }, $$('div', { className: 'coveo-result-cell' }));
+  }
+
+  public getMissingTerms(): string[] {
+    // Keeping only user entered keywords (removing thesaurus or other injected keywords)
+    let cleanMissingTerms = reject(this.result.absentTerms as string[], term => term.indexOf('permanentid') > -1);
```

Comment:
So this function is changed because I know use regex but I still have a need for the casting , i need it because `this.result.absentTerms`by default is of type `any` so i was unable to add it to my array of string. 
To be cleaner, I typed the parameter `term` here's what it looks like now: 
```
public get missingTerms(): string[] {
  let cleanMissingTerms: string[] = []
  each(this.result.absentTerms, (term: string) => {
    const regex = new RegExp(`${this.wordBoundary}${term}${this.wordBoundary}`)
    if (regex.test(this.result.state.q)) {
      cleanMissingTerms.push(term)
    }
  })
  return cleanMissingTerms;
}
```

Reply:
Okay, this is an improvement but I'm interested to see if we could type `this.result.absentTerms` and `this.result` to solve the issue at the source. Otherwise every time someone references `this.result.absentTerms`, they will face the same issue.

For the code using regex, instead of creating a separate array `cleanMissingTerms`, you could use a functional approach using `filter`. e.g.

```
public get missingTerms(): string[] {
    return this.result.absentTerms.filter(term => {
        const regex = new RegExp(`${this.wordBoundary}${term}${this.wordBoundary}`);
        return regex.test(this.result.state.q);
    });
}
```

Reply:
ok after talking with @olamothe I will type the `this.result` as `QueryResult` and add the `absentTerms` property in the interface as an array of string